src/stages/main/patchers/WhilePatcher.js

```diff
@@ -125,9 +126,16 @@ export default class WhilePatcher extends NodePatcher {
   patchAsExpression() {
     this.body.setImplicitlyReturns();
     let resultBinding = this.getResultArrayBinding();
-    this.insert(this.contentStart, `(() => { ${resultBinding} = []; `);
     this.patchAsStatement();
-    this.insert(this.contentEnd, ` return ${resultBinding}; })()`);
+    let prefix = this.yielding ? 'yield* (function*()' : '(() =>';
+    this.insert(this.contentStart, `${prefix} { ${resultBinding} = []; `);
+    let postfix = this.yielding ? '.call(this)' : '()';
```

Comment:
I fixed this in a follow-up commit (https://github.com/decaffeinate/decaffeinate/commit/ac3b6434dc4500ef89c579aaccf1203105b73e26).